components/resource_control/src/service.rs

```diff
@@ -0,0 +1,235 @@
+// Copyright 2023 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::sync::Arc;
+
+use futures::StreamExt;
+use kvproto::{pdpb::ItemKind, resource_manager::ResourceGroup};
+use pd_client::{PdClient, Result, RpcClient};
+use tikv_util::{box_err, error};
+
+use crate::ResourceGroupManager;
+
+pub const CONFIG_PATH: &str = "resource_group/settings";
+
+#[derive(Clone)]
+pub struct ResourceManagerService {
+    pub manager: Arc<ResourceGroupManager>,
+    pd_client: Arc<RpcClient>,
+    // record watch revision
+    pub revision: i64,
+}
+
+impl ResourceManagerService {
+    /// Constructs a new `Service` with `ResourceGroupManager` and a `Rpclient`
+    pub fn new(
+        manager: Arc<ResourceGroupManager>,
+        pd_client: Arc<RpcClient>,
+    ) -> ResourceManagerService {
+        ResourceManagerService {
+            pd_client,
+            manager,
+            revision: 0,
+        }
+    }
+}
+
+impl ResourceManagerService {
+    pub async fn watch_resource_groups(&mut self) {
+        // Firstly, load all resource groups as of now.
+        match self.list_resource_groups() {
+            Ok((groups, revision)) => {
+                self.revision = revision;
+                for group in groups {
+                    self.manager.add_resource_group(group);
+                }
+            }
+            Err(e) => {
+                error!("failed to list resource groups, err: {:?}", e);
+                return;
+            }
+        }
+        // Secondly, start watcher at loading revision.
+        match self
+            .pd_client
+            .watch_global_config(CONFIG_PATH.to_string(), self.revision)
+            .await
+        {
+            Ok(mut stream) => {
+                while let Some(grpc_response) = stream.next().await {
```

Comment:
So what if this stream is stopped by pd due to pd shutdown, the current logic will stop watch but I think it should retry reconnecting to the new pd leader.